core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/CostBasedRegionPlanGenerator.scala

```diff
@@ -194,55 +204,74 @@ class CostBasedRegionPlanGenerator(
     )
 
     while (queue.nonEmpty) {
+      // A state is represented as a set of materialized non-blocking edges.
       val currentState = queue.dequeue()
       visited.add(currentState)
+
       tryConnectRegionDAG(
         physicalPlan.getNonMaterializedBlockingAndDependeeLinks ++ currentState
       ) match {
         case Left(regionDAG) =>
-          // Calculate the current state's cost and update the bestResult if it's lower
-          val cost =
-            evaluate(regionDAG.vertexSet().asScala.toSet, regionDAG.edgeSet().asScala.toSet)
-          if (cost < bestResult.cost) {
-            bestResult = SearchResult(currentState, regionDAG, cost)
-          }
-        // No need to explore further
+          checkSchedulableState(regionDAG)
+          expandFrontier()
         case Right(_) =>
-          val allBlockingEdges =
-            currentState ++ physicalPlan.getNonMaterializedBlockingAndDependeeLinks
-          // Generate and enqueue all neighbour states that haven't been visited
-          val edgesInChainWithBlockingEdge = physicalPlan.maxChains
-            .filter(chain => chain.intersect(allBlockingEdges).nonEmpty)
+          expandFrontier()
+      }
+
+      /**
+        * An internal method of bottom-up search that updates the current optimum if the examined state is schedulable
+        * and has a lower cost.
+        */
+      def checkSchedulableState(regionDAG: DirectedAcyclicGraph[Region, RegionLink]): Unit = {
```

Comment:
the function name does not match the body very much. it is calculating the cost and update the bestResult right? `checkSchedulableState` is not very meaningful.

Reply:
Updated the name